### PR TITLE
mesos net-modules support

### PR DESCRIPTION
**Pull request by unclejack**
This PR adds support for Mesos net-modules.

This PR is WIP.

### CODE REVIEW

**Comment by jainvipin** [comment]
great work!

File: mgmtfn/mesosplugin/api/ipam.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	Hostname  string              `json:"hostname,omitempty"`
+	NumIPv4   int                 `json:"num_ipv4,omitempty"`
+	NumIPv6   int                 `json:"num_ipv6,omitempty"`
+	UID       string              `json:"uid,omitempty"`
+	Netgroups []string            `json:"netgroups,omitempty"`
+	Labels    []map[string]string `json:"labels,omitempty"`
+	IPs       []string            `json:"ips,omitempty"`
+}
+
+type IPAMRequest struct {
```
Comment (shaleman) [inline mgmtfn/mesosplugin/api/ipam.go]: Why do we need this struct? isn't just IPAMRequestArgs sufficient?

File: mgmtfn/mesosplugin/api/ipam.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	IPAMRequestArgs
+}
+
+type IPAMResponse struct {
+	IPV4  []string `json:"ipv4,omitempty"`
+	IPV6  string   `json:"ipv6,omitempty"`
```
Comment (shaleman) [inline mgmtfn/mesosplugin/api/ipam.go]: Why is IPV6 not a slice of strings?

File: mgmtfn/mesosplugin/api/virtualizer.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	Hostname      string              `json:"hostname"`
+	ContainerID   string              `json:"container_id"`
+	PID           int                 `json:"pid,omitempty"`
+	IPV4Addresses []string            `json:"ipv4_addrs,omitempty"`
+	IPV6Addresses []string            `json:"ipv6_addrs,omitempty"`
+	Netgroups     []string            `json:"netgroups,omitempty"`
+	Labels        []map[string]string `json:"labels,omitempty"`
+}
+
+type VirtualizerRequest struct {
```
Comment (shaleman) [inline mgmtfn/mesosplugin/api/virtualizer.go]: Again, i'm not sure why we need the extra struct..

File: mgmtfn/mesosplugin/contiv-isolator/Makefile [hunk 4]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (shaleman) [inline mgmtfn/mesosplugin/contiv-isolator/Makefile]: Why do we need this makefile since we build contiv-isolator as part of top level makefile..

File: mgmtfn/mesosplugin/contiv-isolator/netpluginhandler.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/contiv/netplugin/mgmtfn/mesosplugin/api"
+)
+
+const driverPath = "/run/mesos-netmodules-netplugin.sock"
```
Comment (shaleman) [inline mgmtfn/mesosplugin/contiv-isolator/netpluginhandler.go]: It seems like this belongs in a common place shared between netplugin and contiv-isolator

File: mgmtfn/mesosplugin/api/ipam.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	Error string   `json:"error,omitempty"`
+}
```
Comment (shaleman) [inline mgmtfn/mesosplugin/api/ipam.go]: Technically speaking, "allocate" and "release" messages have different format as defined in https://github.com/mesosphere/net-modules/blob/master/docs/api.md
Can we use different structs for each message?

File: mgmtfn/mesosplugin/api/virtualizer.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	VirtualizerRequestArgs
+}
+
+type VirtualizerResponse struct {
+	Error string `json:"error,omitempty"`
+}
```
Comment (shaleman) [inline mgmtfn/mesosplugin/api/virtualizer.go]: Same comment as above. "isolate" and "cleanup" messages should use different structs since their format is different.. Though they have common fields now, they could differ significantly over time..

File: mgmtfn/mesosplugin/contiv-isolator/netpluginhandler_test.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+*/
+
+package main
+
+/*
+import (
+	"testing"
+)
+
+func TestNetpluginHandler(t *testing.T) {
```
Comment (shaleman) [inline mgmtfn/mesosplugin/contiv-isolator/netpluginhandler_test.go]: Can we please create some unit-test for the netplugin handler in contiv-isolator?
Just to verify its making right HTTP calls to netplugin..

File: mgmtfn/mesosplugin/ipamDriver.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	err = decoder.Decode(&areq)
+	if err != nil {
+		httpError(w, "Could not read and parse requestAddress request", err)
+		return
+	}
+
+	log.Infof("Received RequestAddressRequest: %+v", areq)
+
+	if len(areq.Labels) == 0 {
+		httpError(w, "Labels are missing", fmt.Errorf("Labels length is 0"))
```
Comment (shaleman) [inline mgmtfn/mesosplugin/ipamDriver.go]: I believe labels are optional in mesos. We probably shouldn't error out here

File: mgmtfn/mesosplugin/ipamDriver.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+	var allocResp master.AddressAllocResponse
+	err = cluster.MasterPostReq("/plugin/allocAddress", &allocReq, &allocResp)
+	if err != nil {
+		httpError(w, "master failed to allocate address", err)
+		return
+	}
+
+	addr = strings.Split(allocResp.IPv4Address, "/")[0]
+
+	allocatedEndpoints.Lock()
```
Comment (shaleman) [inline mgmtfn/mesosplugin/ipamDriver.go]: Can we create a separate function for Veth IP allocation?

File: mgmtfn/mesosplugin/ipam_test.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
+/***
```
Comment (shaleman) [inline mgmtfn/mesosplugin/ipam_test.go]: Is this file being used?

File: mgmtfn/mesosplugin/netDriver.go [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+func init() {
+	allocatedEndpoints = &endpoints{
+		epStates: make(map[string]endpointState),
+	}
+	netutils.InitSubnetBitset(&allocatedEndpoints.ipMap, 24)
+	ipAddrValue, _ := allocatedEndpoints.ipMap.NextClear(0)
+	allocatedEndpoints.ipMap.Set(ipAddrValue)
+}
+
+func join(w http.ResponseWriter, r *http.Request) {
```
Comment (shaleman) [inline mgmtfn/mesosplugin/netDriver.go]: Can we name this "isolate" to reflect mesos terminology?

File: mgmtfn/mesosplugin/netDriver.go [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+type endpointState struct {
+	endpointRequest master.CreateEndpointRequest
+	endpoint        *drivers.OvsOperEndpointState
+	network         *mastercfg.CfgNetworkState
+	vethMasterName  string
+	vethIP          string
+}
+
+type endpoints struct {
+	epStates map[string]endpointState
```
Comment (shaleman) [inline mgmtfn/mesosplugin/netDriver.go]: @unclejack Can you please open a github issue to persist this state? It'll be easy to keep track that way..

File: mgmtfn/mesosplugin/netDriver.go [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+			GatewayIP: "10.112.95.1",
+		},
+		{
+			Name: "lo",
+		},
+	}
+
+	log.Infof("nics: %+v", nics)
+	/*
+		out, err := exec.Command("/sbin/ifconfig").CombinedOutput()
```
Comment (shaleman) [inline mgmtfn/mesosplugin/netDriver.go]: Cleanup unused code :)

File: mgmtfn/mesosplugin/netDriver.go [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+	content, err = json.Marshal(epResponse)
+	if err != nil {
+		httpError(w, "Could not generate create endpoint response", err)
+		return
+	}
+
+	w.Write(content)
+}
+
+func leave(w http.ResponseWriter, r *http.Request) {
```
Comment (shaleman) [inline mgmtfn/mesosplugin/netDriver.go]: Same comment as above:  Name it as "cleanup" to reflect mesos terminology

File: mgmtfn/mesosplugin/netDriver.go [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+	content, err = json.Marshal(leaveResp)
+	if err != nil {
+		httpError(w, "Could not generate leave response", err)
+		return
+	}
+
+	w.Write(content)
+}
+
+func netdGetEndpoint(epID string) (*drivers.OvsOperEndpointState, error) {
```
Comment (shaleman) [inline mgmtfn/mesosplugin/netDriver.go]: Can we move this to a common package since both dockplugin and mesosplugin need it?

File: netlib/network.go [hunk 17]
```diff
@@ -0,0 +1,288 @@
+package netlib
```
Comment (shaleman) [inline netlib/network.go]: Can we please move this file to utils/netlib directory instead of creating top level netlib directory?

File: vagrant/mesos-netmodules/Vagrantfile [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+sudo systemctl stop docker
+systemctl start docker-tcp.socket
+sudo systemctl start docker
+
+if [ $# -gt 7 ]; then
+    shift; shift; shift; shift; shift; shifti; shift
+    echo "export $@" >> /etc/profile.d/envvar.sh
+fi
+# Get swarm binary
+# (wget https://cisco.box.com/shared/static/0txiq5h7282hraujk09eleoevptd5jpl -q -O /usr/bin/swarm &&
```
Comment (shaleman) [inline vagrant/mesos-netmodules/Vagrantfile]: We don't need swarm binary on these VMs..

File: vagrant/mesos-netmodules/Vagrantfile [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+echo "export USE_RELEASE=$6" >> /etc/profile.d/envvar.sh
+echo "export no_proxy=$7,127.0.0.1,localhost,netmaster" >> /etc/profile.d/envvar.sh
+echo "export CLUSTER_NODE_IPS=$7" >> /etc/profile.d/envvar.sh
+source /etc/profile.d/envvar.sh
+mv /etc/resolv.conf /etc/resolv.conf.bak
+cp #{gopath_folder}/src/github.com/contiv/netplugin/resolv.conf /etc/resolv.conf
+# setup docker cluster store
+cp #{gopath_folder}/src/github.com/contiv/netplugin/scripts/docker.service /lib/systemd/system/docker.service
+# setup docker remote api
+cp #{gopath_folder}/src/github.com/contiv/netplugin/scripts/docker-tcp.socket /etc/systemd/system/docker-tcp.socket
```
Comment (shaleman) [inline vagrant/mesos-netmodules/Vagrantfile]: I don't think you need to enable docker remote API for mesos

File: vagrant/mesos-netmodules/Vagrantfile [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+echo "export http_proxy='$4'" >> /etc/profile.d/envvar.sh
+echo "export https_proxy='$5'" >> /etc/profile.d/envvar.sh
+echo "export USE_RELEASE=$6" >> /etc/profile.d/envvar.sh
+echo "export no_proxy=$7,127.0.0.1,localhost,netmaster" >> /etc/profile.d/envvar.sh
+echo "export CLUSTER_NODE_IPS=$7" >> /etc/profile.d/envvar.sh
+source /etc/profile.d/envvar.sh
+mv /etc/resolv.conf /etc/resolv.conf.bak
+cp #{gopath_folder}/src/github.com/contiv/netplugin/resolv.conf /etc/resolv.conf
+# setup docker cluster store
+cp #{gopath_folder}/src/github.com/contiv/netplugin/scripts/docker.service /lib/systemd/system/docker.service
```
Comment (shaleman) [inline vagrant/mesos-netmodules/Vagrantfile]: We don't need to setup docker cluster store for mesos

File: vagrant/mesos-netmodules/Vagrantfile [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+# (wget https://cisco.box.com/shared/static/0txiq5h7282hraujk09eleoevptd5jpl -q -O /usr/bin/swarm &&
+# chmod +x /usr/bin/swarm) || exit 1
+#Get gobgp binary
+wget https://cisco.box.com/shared/static/5leqlo84kjh0thty91ouotilm4ish3nz -q -O #{gopath_folder}/src/github.com/contiv/netplugin/scripts/gobgp && chmod +x #{gopath_folder}/src/github.com/contiv/netplugin/scripts/gobgp 
+# remove duplicate docker key
+rm /etc/docker/key.json
+(service docker restart) || exit 1
+(ovs-vsctl set-manager tcp:127.0.0.1:6640 && \
+ ovs-vsctl set-manager ptcp:6640) || exit 1
+docker load --input #{gopath_folder}/src/github.com/contiv/netplugin/scripts/dnscontainer.tar
```
Comment (shaleman) [inline vagrant/mesos-netmodules/Vagrantfile]: Are you planning to make use of skydns container in mesos?

File: vagrant/mesos-netmodules/Vagrantfile [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+ --listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001 \
+ --advertise-client-urls http://#{node_addr}:2379,http://#{node_addr}:4001 \
+ --initial-advertise-peer-urls http://#{node_addr}:2380,http://#{node_addr}:7001 \
+ --listen-peer-urls http://#{node_addr}:2380 \
+ --initial-cluster #{node_peers.join(",")} --initial-cluster-state new \
+  0<&- &>/tmp/etcd.log &) || exit 1
+## start consul
+(nohup consul agent -server #{consul_join_flag} #{consul_bootstrap_flag} \
+ -bind=#{node_addr} -data-dir /opt/consul 0<&- &>/tmp/consul.log &) || exit 1
+# start swarm
```
Comment (shaleman) [inline vagrant/mesos-netmodules/Vagrantfile]: no need to start swarm in mesos cluster

File: vagrant/mesos-netmodules/mesos-setup.sh [hunk 23]
```diff
@@ -0,0 +1,201 @@
+#!/bin/bash
+
```
Comment (shaleman) [inline vagrant/mesos-netmodules/mesos-setup.sh]: Can we create a vbox image with all this pre-installed?
Its perfectly fine to do that as a separate PR

**Comment by shaleman** [comment]
@unclejack 
Overall good work. Thanks for pulling this in.

Couple of high level comments:
1. Please create a Readme in mgmtfn/mesos-plugin directory on how to try mesos plugin. I.e, where to do "vagrant up" what commands to issue inside the vagrant VMs to launch Marathon jobs on contiv network etc..
2. Along the same lines, it'll be good to add "make mesos-demo" and "make mesos-destroy" make targets into Makefile.
3. It'll be good to create a box image with all mesos packages pre-installed so that it'll be easier for people to try. This is a lower priority and can be done separately..

File: mgmtfn/mesosplugin/ipam_test.go [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+	tr := &http.Transport{
+		Dial: fakeDial,
+	}
+	client = &http.Client{Transport: tr}
+}
+
+func fakeDial(proto, addr string) (conn net.Conn, err error) {
+	fmt.Printf("dialing with proto %v and address %v", proto, addr)
+	return net.DialTimeout("unix", driverPath, time.Second*5)
+}
```
Comment (shaleman) [inline mgmtfn/mesosplugin/ipam_test.go]: Can we create a unit-test of some sort for mesos-plugin ?

**Comment by shaleman** [comment]
Mesos CNI plugin changes have been merged. Closing this PR for now..
